File: cluster_analysis.cpp
/*
  Copyright (C) 2011,2012,2013,2014,2015,2016 The ESPResSo project
  
  This file is part of ESPResSo.
  
  ESPResSo is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/
#include "cluster_analysis.hpp"
#include "particle_data.hpp"
#include "collision.cpp"
#include "collision.hpp"	
#ifdef COLLISION_DETECTION


void ClusterStructure::clear() {
 clusters.clear();
 cluster_id.clear();
 cluster_identities.clear();
}


// Analyze the cluster structure of the given particles
void ClusterStructure::analyze_pair(Particle* p1, Particle* p2)
{
  // clear data structs
  clear();
  
  // Iterate over all possible pairs (n=i*j particles) - ghosts are not included
  for (int i=0;i<=max_seen_particle;i++) {
    if (! local_particles[i]) continue;
    for (int j=i+1;j<=max_seen_particle;j++) {
      if (! local_particles[j]) continue;
      count(*local_particles[i],*local_particles[j]); // maybe no * 
//      int part1, part2;
//      part1 = p1->p.identity;
//      part2 = p2->p.identity;
//      p1 = local_particle[part1];
//      p2 = local_particle[part2];      
//      //check if bond exist for the two particles
//      if (!(p1==p2)) {
//        if ((bond_exists(p1,p2, collision_params.bond_centers)) or (bond_exists(p2,p1, collision_params.bond_centers))){
          //there exist bond, which means that particles are neighbors 
//          return true;
//          printf("particles %d and %d are neighbors!\n",p1.p.identity, p2.p.identity);
        } //bond exist
        else 
          return false;
      } //particles are not the same


    }
  }
}


void ClusterStructure::add_pair(Particle& p1, Particle& p2) {
// * check, if there's a neighbor
 //   * No: Then go on to the next particle
 // * Yes: Then if
 //   * One of them belongs to a cluster, give the other one the same cluster
 //     id.
 //   * None of them belongs to a cluster: Give them both a new cluster id
 //   * Both belong to different clusters: Mark the clusters as identical 
 //   * so that they can be put together later
  if (! nc) {
  //if (! nc) {
    runtimeErrorMsg() << "No cluster criterion defined"; 
    return;
  }
  if (nc->are_neighbors(p1,p2)) {
     if // None belongs to a cluster
     ((cluster_id.find(p1->p.identity)==cluster_id.end()) && (cluster_id.find(p2->p.identity)==cluster_id.end())
    {  
      // Both particles belong to the same, new cluster
      cid=get_next_free_cluster_id();
      cluster_id[p1->p.identity]=cid;
      cluster_id[p2->p.identity]=cid;
    }
    else if // j belongs to a cluster but i doesn't
      ((cluster_id.find(p1.p.identity)==cluster_id.end()) 
      && 
      (cluster_id[p2.p.identity] !=cluster_id.end()))
    {
     // Give p1 the same cluster id as p2
     cluster_id[p1.p.identity]=find_id_for(cluster_id[p2.p.identity]);
    }
    else if // i belongs to a cluster but j doesn't: 
      ((cluster_id.find(p2.p.identity)==cluster_id.end()) 
      && 
      (cluster_id[p1.p.identity] !=cluster_id.end()))
    {
     // give p2 the cluster id from p1
     cluster_id[p2.p.identity]=find_id_for(cluster_id[p1.p.identity]);
    }
    else if // Both belong to different clusters - merge clusters
      (cluster_id[p1.p.identity] != cluster_id[p2.p.identity])
    {
     // Clusters of p1 and p2 are one and the same. Add an identity to the list
     // The lower number must be inserted as first value of tjhe pair
     // because the substituions later have to be done in ascending order
     if (cluster_id[p1.p.identity]<cluster_id[p2.p.identity])
     {
       cluster_identities[find_id_for(cluster_id[p2.p.identity])] =find_id_for(cluster_id[p1.p.identity]);
     }
     else
     {
       cluster_identities[find_id_for(cluster_id[p1.p.identity])] =find_id_for(cluster_id[p2.p.identity]);
     }
     
     // Connected clusters will be merged later
    }
    // The case for both particles being in the same cluster does not need to be
    // treated.
  }
}


void ClusterStructure::merge_clusters() {
  // Relabel particles according to the cluster identities map
  // Also create empty cluster objects for the final cluster id
  for (auto it : cluster_ids) { 
    // particle id is in it.first and cluster id in it.second
    // We change the cluster id according to the cluster identities
    // map
    cid=find_id_for(it.second);
    it.second =cid;
    // Empty cluster object
    if (clusters.find(cid)==clusters.end()) {
      clusters[cid]=Cluster();
    }
  }
  // Now fill the cluster objects with particle ids
  // Itterate over particles
  for (auto id : cluster_ids) {
    clusters[it.second].particles.push_back(it.first);
  }
}


int ClusterStructure::find_id_for(int x)
{
 while (cluster_identities.find(x)!=cluster_identities.end())
 {
  x =cluster_identities[x];
 }
 return x;
}


//ClusterStructure cluster_structure;
ClusterStructure& cluster_analysis() {
  return cluster_structure;
}


void ClusterStructure::set_criterion(NeighborCriterion c) {
  if (nc)
  {
    delete nc;
    nc=0;
  }
  nc=c;
}



#endif   

 

File: src/python/espressomd/cluster_analysis.py
# Copyright (C) 2010-2019 The ESPResSo project
#
# This file is part of ESPResSo.
#
# ESPResSo is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ESPResSo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
from .script_interface import ScriptInterfaceHelper, script_interface_register
from .particle_data import ParticleHandle, ParticleSlice


@script_interface_register
class Cluster(ScriptInterfaceHelper):

    """Class representing a cluster of particles.

    Methods
    -------
    particle_ids()
        Returns list of particle ids in the cluster

    size()
        Returns the number of particles in the cluster

    center_of_mass()
        Center of mass of the cluster

    longest_distance()
        Longest distance between any combination of two particles in the cluster

    fractal_dimension(dr=None)
        Estimates the cluster's fractal dimension by fitting the number of
        particles :math:`n` in spheres of growing radius around the center of mass
        to :math:`c*r_g^d`, where :math:`r_g` is the radius of gyration of the
        particles within the sphere, and :math:`d` is the fractal dimension.

        .. note::

            Requires ``GSL`` external feature, enabled with ``-DWITH_GSL=ON``.

        Parameters
        ----------
        dr: :obj:`float`
            Minimum increment for the radius of the spheres.

        Returns
        -------
        :obj:`tuple`:
            Fractal dimension and mean square residual.

    """
    _so_name = "ClusterAnalysis::Cluster"
    _so_bind_methods = ("particle_ids", "size", "longest_distance",
                        "radius_of_gyration", "fractal_dimension", "center_of_mass")

    _so_creation_policy = "LOCAL"

    def particles(self):
        """
        Get particles in the cluster.

        Returns
        -------
        :class:`espressomd.particle_data.ParticleSlice`

        """
        return ParticleSlice(self.particle_ids())


@script_interface_register
class ClusterStructure(ScriptInterfaceHelper):

    """Cluster structure of a simulation system, and access to cluster analysis

    Parameters
    ----------
    pair_criterion: :class:`espressomd.pair_criteria._PairCriterion`
        Criterion to decide whether two particles are neighbors.

    """
    _so_name = "ClusterAnalysis::ClusterStructure"
    _so_creation_policy = "LOCAL"

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._clusters = Clusters(self)

    def run_for_all_pairs(self):
        """
        Runs the cluster analysis, considering all pairs of particles in the system

        """
        return self.call_method("run_for_all_pairs")

    def run_for_bonded_particles(self):
        """
        Runs the cluster analysis, considering only pairs of particles connected by a pair-bond.

        """
        return self.call_method("run_for_bonded_particles")

    def clear(self):
        """
        Clears the cluster structure.

        """
        return self.call_method("clear")

    def cluster_ids(self):
        """
        Returns a list of all cluster ids of the clusters in the structure.

        """
        return self.call_method("cluster_ids")

    def cid_for_particle(self, p):
        """Returns cluster id for the particle.

        Parameters
        ----------
        p : :obj:`espressomd.particle_data.ParticleHandle` or :obj:`int` containing the particle id
            Particle.

        """
        if isinstance(p, ParticleHandle):
            return self.call_method("cid_for_particle", pid=p.id)
        if isinstance(p, int):
            return self.call_method("cid_for_particle", pid=p)
        raise TypeError(
            "The particle has to be passed as instance of ParticleHandle or as an integer particle id")

    @property
    def clusters(self):
        """Gives access to the clusters in the cluster structure via an
        instance of :class:`Clusters`."""
        return self._clusters


class Clusters:

    """Access to the clusters in the cluster structure.

    Access is as follows:

    * number of clusters: ``len(clusters)``
    * access a cluster via its id: ``clusters[id]``
    * iterate over clusters::

          for c in clusters:

      where ``c`` will be a tuple containing the cluster id and the cluster object.

    """

    def __init__(self, cluster_structure):
        self.cluster_structure = cluster_structure

    def __getitem__(self, cluster_id):
        return self.cluster_structure.call_method("get_cluster", id=cluster_id)

    def __iter__(self):
        for cid in self.cluster_structure.cluster_ids():
            yield (cid, self.cluster_structure.call_method("get_cluster", id=cid))

    def __len__(self):
        return self.cluster_structure.call_method("n_clusters")
